proxy/proxy.go

```diff
@@ -0,0 +1,166 @@
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"sync"
+)
+
+// Main struct holding the proxy state
+type proxy struct {
+	// Protect concurrent accesses from separate client goroutines to this
+	// structure fields
+	sync.Mutex
+
+	// client socket
+	listener net.Listener
+
+	// vms are hashed by their containerID
+	vms map[string]*vm
+}
+
+// Represents a client, either a cc-oci-runtime or cc-shim process having
+// opened a socket to the proxy
+type client struct {
+	proxy *proxy
+	vm    *vm
+
+	conn net.Conn
+}
+
+// "hello"
+type helloCmd struct {
+	ContainerId string `json:containerId`
+	CtlSerial   string `json:ctlSerial`
+	IoSerial    string `json:ioSerial`
+}
+
+func helloHandler(data []byte, userData interface{}) (map[string]interface{}, error) {
+	client := userData.(*client)
+	hello := helloCmd{}
+
+	if err := json.Unmarshal(data, &hello); err != nil {
+		return nil, err
+	}
+
+	if hello.ContainerId == "" || hello.CtlSerial == "" || hello.IoSerial == "" {
+		return nil, errors.New("malformed hello command")
+	}
+
+	proxy := client.proxy
+	proxy.Lock()
+	if _, ok := proxy.vms[hello.ContainerId]; ok {
+
+		proxy.Unlock()
+		return nil, errors.New("container already registered")
+	}
+	vm := NewVM(hello.ContainerId, hello.CtlSerial, hello.IoSerial)
+	client.vm = vm
+	proxy.vms[hello.ContainerId] = vm
+	proxy.Unlock()
+
+	if err := vm.Connect(); err != nil {
+		proxy.Lock()
+		delete(proxy.vms, hello.ContainerId)
+		proxy.Unlock()
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+// "hyper"
+type hyperCmd struct {
+	Name string          `json:name`
+	Data json.RawMessage `json:data`
+}
+
+func hyperHandler(data []byte, userData interface{}) (map[string]interface{}, error) {
+	client := userData.(*client)
+	hyper := hyperCmd{}
+	vm := client.vm
+
+	if err := json.Unmarshal(data, &hyper); err != nil {
+		return nil, err
+	}
+
+	if vm == nil {
+		return nil, errors.New("client not attached to a vm")
+	}
+
+	err := vm.SendMessage(hyper.Name, hyper.Data)
+	return nil, err
+}
+
+func NewProxy() *proxy {
+	return &proxy{
+		vms: make(map[string]*vm),
+	}
+}
+
+func (proxy *proxy) Init() error {
+	fds := listenFds()
+
+	if len(fds) != 1 {
+		return fmt.Errorf("couldn't find activated socket (%d)", len(fds))
+	}
+
+	fd := fds[0]
+	l, err := net.FileListener(fd)
+	if err != nil {
+		return fmt.Errorf("couldn't listen on socket: %s", err.Error())
+	}
+
+	proxy.listener = l
+
+	return nil
+}
+
+func (proxy *proxy) Serve() {
+
+	// Define the client (runtime/shim) <-> proxy protocol
+	proto := NewJsonProto()
+	proto.Handle("hello", helloHandler)
+	proto.Handle("hyper", hyperHandler)
+
+	for {
+		conn, err := proxy.listener.Accept()
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "couldn't accept connection:", err)
+			continue
+		}
+
+		newClient := &client{
+			proxy: proxy,
+			conn:  conn,
+		}
+
+		go proto.Serve(conn, newClient)
+	}
+}
+
```

Comment:
A global `stop()` doesn't seem like it'd fit the model. Each client has (now) a hello()/bye(), but globally noone can tell to proxy to "stop()" because they only have local knowledge of their own qemu lifetime.

What would be possible though, is to exit the proxy once we've lost all clients (or even do that after a time without new clients), a new client would start the proxy again through socket activation.